gadopt/diagnostics.py

```diff
@@ -35,37 +39,54 @@ def __init__(
         T: Function,
         bottom_id: int,
         top_id: int,
+        diag_vars: dict = {},
         degree: int = 4
     ):
-        self.mesh = extract_unique_domain(u)
+        mesh = extract_unique_domain(u)
+
         self.u = u
         self.p = p
         self.T = T
+        self.diag_vars = diag_vars
 
-        self.dx = firedrake.dx(degree=degree)
-        if T.function_space().extruded:
-            self.ds = CombinedSurfaceMeasure(self.mesh, degree)
-        else:
-            self.ds = firedrake.ds(self.mesh)
+        self.dx = dx(domain=mesh, degree=degree)
+        self.ds = (
+            CombinedSurfaceMeasure(mesh, degree)
+            if T.function_space().extruded
+            else ds(mesh)
+        )
         self.ds_t = self.ds(top_id)
         self.ds_b = self.ds(bottom_id)
 
-        self.n = FacetNormal(self.mesh)
+        self.n = FacetNormal(mesh)
 
-    def domain_volume(self) -> float:
-        return assemble(Constant(1)*firedrake.dx(domain=self.mesh))
+        self.domain_volume = assemble(Constant(1) * self.dx)
+        self.top_surface = assemble(Constant(1) * self.ds_t)
+        self.bottom_surface = assemble(Constant(1) * self.ds_b)
 
-    def u_rms(self) -> float:
-        return sqrt(assemble(dot(self.u, self.u) * self.dx)) * sqrt(1./self.domain_volume())
+    def u_rms(self):
+        return norm(self.u) / sqrt(self.domain_volume)
 
     def u_rms_top(self) -> float:
         return sqrt(assemble(dot(self.u, self.u) * self.ds_t))
 
-    def Nu_top(self) -> float:
-        return -1 * assemble(dot(grad(self.T), self.n) * self.ds_t) * (1./assemble(Constant(1) * self.ds_t))
+    def Nu_top(self):
+        return -assemble(dot(grad(self.T), self.n) * self.ds_t) / self.top_surface
+
+    def Nu_bottom(self):
+        return assemble(dot(grad(self.T), self.n) * self.ds_b) / self.bottom_surface
+
+    def T_avg(self):
+        return assemble(self.T * self.dx) / self.domain_volume
+
+    def entrainment(self, level_set_index, material_area, entrainment_height):
+        level_set = self.diag_vars["level_set"][level_set_index]
 
-    def Nu_bottom(self) -> float:
-        return assemble(dot(grad(self.T), self.n) * self.ds_b) * (1./assemble(Constant(1) * self.ds_b))
+        mesh_coords = SpatialCoordinate(level_set.function_space().mesh())
+        target_region = mesh_coords[1] >= entrainment_height
+        material_entrained = conditional(level_set < 0.5, 1, 0)
 
-    def T_avg(self) -> float:
-        return assemble(self.T * self.dx) / self.domain_volume()
+        return (
+            assemble(conditional(target_region, material_entrained, 0) * self.dx)
+            / material_area
+        )
```

Comment:
Yeah, I would just make this a separate function. Tbh I can't quite remember why we thought combining these other basic diagnostics in one class was a good idea, maybe the idea was to have it as a diagnostic _file_ class originally? But in any case this one in particular seems quite separated from the class as a whole

Reply:
Well, I find it pretty neat that you can just instantiate a class by providing velocity, pressure, temperature, and boundary IDs and then just query diagnostics as methods. I think it is a good idea. Regarding `entrainment`, it is a very common diagnostic in multi-material geodynamical simulations. Given the class is called `GeodynamicalDiagnostics`, perhaps it should be included as a method. But it indeed requires arguments, whereas other existing methods do not. So it is not as straightforward as the other ones. @drhodrid @sghelichkhani Any opinions?